base/compiler/optimize.jl

```diff
@@ -28,14 +28,14 @@ struct InliningState{S <: Union{EdgeTracker, Nothing}, T, P}
     policy::P
 end
 
-function default_inlining_policy(@nospecialize(src))
+function default_inlining_policy(@nospecialize(src), stmt_flag::Union{Nothing,UInt8})
     if isa(src, CodeInfo) || isa(src, Vector{UInt8})
         src_inferred = ccall(:jl_ir_flag_inferred, Bool, (Any,), src)
-        src_inlineable = ccall(:jl_ir_flag_inlineable, Bool, (Any,), src)
+        src_inlineable = is_stmt_inline(stmt_flag) || ccall(:jl_ir_flag_inlineable, Bool, (Any,), src)
         return src_inferred && src_inlineable ? src : nothing
     end
     if isa(src, OptimizationState) && isdefined(src, :ir)
-        return src.src.inlineable ? src.ir : nothing
+        return (is_stmt_inline(stmt_flag) || src.src.inlineable) ? src.ir : nothing
```

Comment:
I think the easy and customizable approach to support callsite `@inline` annotation is to check the statement flag within `inlining_policy`, and thus I ended up changing the signature of the interface.
@Keno does it sound reasonable/okay to you ?